### ISSUE TITLE
[Feature Request] Switch and reproduce label-aware augmentation policies with Hydra multirun

### ISSUE THREAD

**Issue by ternaus** [facebookresearch/hydra#3419]
# 🚀 Feature Request

Would you be open to a small real-world example for computer-vision users who need to switch complete, label-aware augmentation policies between experiments without editing Python code, then reproduce one transformed sample from a selected Hydra job?

## User problem

An augmentation policy is not a single scalar parameter. It is a nested object graph containing transforms, probabilities, and annotation-processing rules. When that graph is constructed directly in training code, switching between `light` and `strong` policies requires code changes or application-specific branching, and the selected policy is harder to sweep and inspect as one experiment input.

Hydra config groups and multirun can select and compare complete policies. AlbumentationsX can construct the corresponding label-aware pipeline, serialize the resulting policy, and apply it to an image, bounding box, and class label with a stable sample-level `invocation_seed`.

Together, the example would let a user:

- switch the whole augmentation policy through a Hydra override instead of changing training code;
- sweep policies and sample seeds with multirun;
- inspect both the Hydra config and the resolved AX policy produced by each job; and
- reconstruct the policy and reproduce the same transformed labeled sample from a selected job.

This last step exposes an important reproducibility boundary. A composed Hydra config records which policy and overrides a job used, while reproducing one stochastic preprocessing result additionally needs:

1. the resolved runtime policy after object construction; and
2. a stable sample-level invocation seed.

## Why this fits Hydra

Hydra's current object-instantiation documentation explains nested targets with intentionally simple optimizer, dataset, and trainer classes. A computer-vision augmentation policy provides the same structure in a compact real-world workflow: a composition owns a list of transform objects plus annotation-processing parameters, and complete alternatives can be represented as config groups.

The example would complement the current work on Hydra rerun and execution reproducibility in #3248 and #3326. It would not claim to pin source code or the complete runtime environment; it would demonstrate the application-level artifacts needed to select, compare, and reproduce stochastic preprocessing.

## Pitch

Add one self-contained example, with placement left to the maintainers, in which a user compares two policies on the same generated image, bounding box, and class label and verifies that a selected job can reconstruct its result. Generated data keeps the example focused on policy selection and reproduction rather than dataset download or model training.

The example would contain two augmentation config-group options, such as `light` and `strong`. To make the configuration boundary explicit, the `light` option would be equivalent to constructing this pipeline directly in Python:

```python
import albumentations as A

augmentation = A.Compose(
    [
        A.HorizontalFlip(p=0.5),
        A.RandomBrightnessContrast(p=0.2),
    ],
    bbox_params=A.BboxParams(
        coord_format="pascal_voc",
        label_fields=["class_labels"],
    ),
)
```

The same object graph would be expressed as a Hydra config group:

```yaml
# conf/augmentation/light.yaml
_target_: albumentations.Compose
transforms:
  - _target_: albumentations.HorizontalFlip
    p: 0.5
  - _target_: albumentations.RandomBrightnessContrast
    p: 0.2
bbox_params:
  _target_: albumentations.BboxParams
  coord_format: pascal_voc
  label_fields:
    - class_labels
```

with the root config supplying the selected group and a sample-level seed:

```yaml
# conf/config.yaml
defaults:
  - augmentation: light
  - _self_

sample_seed: 17
```

The application would instantiate that graph under the current narrow target-authorization API. The resulting `augmentation` object is equivalent to the direct Python construction above:

```python
from hydra.utils import instantiate, target_whitelist

with target_whitelist("albumentations.*"):
    augmentation = instantiate(cfg.augmentation)
```

A multirun could then compare complete policies and seeds:

```console
python app.py -m augmentation=light,strong sample_seed=17,29
```

For each job, the example would:

- apply the policy to the same generated image, bounding box, and class label with `invocation_seed=cfg.sample_seed`;
- save the fully resolved AX policy with `A.to_dict()`;
- restore it with `A.from_dict()`;
- reapply the restored policy to the same input with the same `invocation_seed` and verify that the same transformed image, box, and label are produced.

Together, the job output would show three distinct layers:

```text
Hydra config + overrides -> resolved AX policy -> sample-level invocation seed
```

This would require no Hydra plugin, launcher, sweeper, or core API change. It could live under `examples/instantiate`, or remain an external integration recipe if project-specific dependencies are not appropriate for the Hydra repository.

### Alternatives considered

- Constructing the augmentation directly in Python omits Hydra config groups, overrides, and multirun.
- Saving only the Hydra config records the policy definition but not the sample-level seed needed to reproduce one stochastic result independently of worker ordering.
- `ReplayCompose` can preserve the exact sampled parameter payload for forensic debugging, but it is not required for this base example. A normal `Compose` plus `invocation_seed` better represents the common training path.
- Publishing only in the AlbumentationsX documentation would still help AX users, but it would not provide Hydra users with a tested real-world nested-instantiation example. I am happy to use that route if maintainers prefer not to carry the dependency.

### Dependency boundary

- The public `albumentationsx` package is AGPL-3.0-only and imports as `albumentations`.
- Importing it requires PyTorch. PyTorch is intentionally not pinned in the package metadata because users need to choose the appropriate CPU, CUDA, or MPS build.
- Both dependencies would remain scoped to this example and would not become Hydra runtime dependencies.

**Are you willing to open a pull request?**

Yes. I would wait for maintainer feedback on scope and placement before preparing it.

## Additional context

I searched the current repository, Issues, and pull requests for `Albumentations`, `AlbumentationsX`, and `albumentationsx` and found no existing integration or proposal.

**Comment by omry** [comment]
Hi @ternaus,
Hydra cannot maintain documentation and usage patterns for downstream libraries. You’re welcome to add this example to the downstream project and link to the relevant Hydra documentation from there.

Thanks for the proposal. I’m closing this as not planned.
